### ISSUE TITLE
Fix ImplementIDisposableCorrectly: Rule should not fail on partial classes

### ISSUE THREAD

**Issue by nnpcYvIVl** [SonarSource/sonar-csharp#416]
### Description
This error is thrown during analysis: AD0001 Analyzer 'SonarAnalyzer.Rules.CSharp.ImplementIDisposableCorrectly' threw an exception of type 'System.ArgumentException' with message 'Syntax node is not within syntax tree'.

### Related information

* SonarC# Latest Version
* Visual Studio 2017

**Comment by Evangelink** [comment]
Hi @nnpcYvIVl!

Could you manage to create some small reproducer? Or at worst to share with us the class where this happens (you can send us a direct email if you are concerned about the privacy of your code).

Cheers,
Amaury

**Comment by nnpcYvIVl** [comment]
It's a large project and I haven't been able to create a small repro. I'll keep trying but for now here is the stack trace.

```
Severity	Code	Description	Project	File	Line	Category	Suppression State	Detail Description
Warning	AD0001	Analyzer 'SonarAnalyzer.Rules.CSharp.ImplementIDisposableCorrectly' threw an exception of type 'System.ArgumentException' with message 'Syntax node is not within syntax tree'.	CCM		1	Compiler	Active	Analyzer 'SonarAnalyzer.Rules.CSharp.ImplementIDisposableCorrectly' threw the following exception:
'Exception occurred with following context:
Compilation: Redacted
SyntaxTree: Redacted\...\Redacted.cs
SyntaxNode: public partial class Redacted ... [ClassDeclarationSyntax]@[87..967) (6,4)-(34,5)

System.ArgumentException: Syntax node is not within syntax tree
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.CheckSyntaxNode(CSharpSyntaxNode syntax)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetSymbolInfo(ExpressionSyntax expression, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpExtensions.GetSymbolInfo(SemanticModel semanticModel, ExpressionSyntax expression, CancellationToken cancellationToken)
   at SonarAnalyzer.Helpers.SyntaxHelper.<>c__DisplayClass15_0.<ContainsMethodInvocation>b__0(InvocationExpressionSyntax e)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.<OfTypeIterator>d__92`1.MoveNext()
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at SonarAnalyzer.Helpers.SyntaxHelper.ContainsMethodInvocation(IEnumerable`1 syntaxNodes, SemanticModel semanticModel, Func`2 syntaxPredicate, Func`2 symbolPredicate)
   at SonarAnalyzer.Helpers.SyntaxHelper.ContainsMethodInvocation(BaseMethodDeclarationSyntax methodDeclarationBase, SemanticModel semanticModel, Func`2 syntaxPredicate, Func`2 symbolPredicate)
   at SonarAnalyzer.Rules.CSharp.ImplementIDisposableCorrectly.DisposableChecker.CallsVirtualDispose(BaseMethodDeclarationSyntax methodDeclaration, String argumentValue)
   at SonarAnalyzer.Rules.CSharp.ImplementIDisposableCorrectly.DisposableChecker.VerifyDisposeOverride(MethodDeclarationSyntax disposeMethod)
   at SonarAnalyzer.Rules.CSharp.ImplementIDisposableCorrectly.DisposableChecker.GetIssueLocations()
   at SonarAnalyzer.Rules.CSharp.ImplementIDisposableCorrectly.<>c.<Initialize>b__6_0(SyntaxNodeAnalysisContext c)
   at SonarAnalyzer.Helpers.DiagnosticAnalyzerContextHelper.<>c__DisplayClass0_0`1.<RegisterSyntaxNodeActionInNonGenerated>b__0(SyntaxNodeAnalysisContext c)
   at SonarAnalyzer.Helpers.SonarAnalysisContext.<>c__DisplayClass12_0`1.<RegisterSyntaxNodeAction>b__0(SyntaxNodeAnalysisContext c)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c__DisplayClass42_1`1.<ExecuteSyntaxNodeAction>b__1()
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock(DiagnosticAnalyzer analyzer, Action analyze, Nullable`1 info)
-----
'.
```

**Comment by Evangelink** [comment]
Hi @nnpcYvIVl, 
thanks for this stacktrace, we will have a look tomorrow and see whether we can find what's wrong.

**Comment by Evangelink** [comment]
Hi again @nnpcYvIVl,
The problem seems to be linked to the `Redacted` class which seems to be a partial implementation. I haven't find a way to reproduce the error yet. Would you mind sharing with me this class? (amaury[dot]leve[@]sonarsource[dot]com).
In case you can't share it, could you try to create reproducer based on the setup of this class (we need only the `IDisposable` related things like: Dispose methods, finalizers, if the class implements `IDiposable`, if the class is partial...

**Comment by nnpcYvIVl** [comment]
Repro code sent to your email.

**Comment by Evangelink** [comment]
Hi @nnpcYvIVl,

I still cannot reproduce the error with your code example. I have tried with VS2015 and VS2017. Could you tell me the exact steps using your repro case to have the issue? Also which version of the SonarLint for Visual Studio are you using?

**Comment by nnpcYvIVl** [comment]
To reproduce the warning simply open the solution file and within a second or two it will show in the error list.

I'm not sure if this matters but you may want to check the following setting "Settings/Text Editor/C#/Advanced/Enable full solution analysis". It is not on by default but I normally keep it enabled.

Here is my version information: 

Microsoft Visual Studio Enterprise 2017 
Version 15.2 (26430.14) Release
VisualStudio.15.Release/15.2.0+26430.14
Microsoft .NET Framework
Version 4.6.01590

SonarLint for Visual Studio   3.1.1.1583

**Comment by Evangelink** [comment]
@nnpcYvIVl Issue reproduced and will be fixed for next release (6.3). 

Thanks for the feedback!

**Comment by nnpcYvIVl** [comment]
Thanks!
